### PR TITLE
Flickable: Shift + mouse scroll invert scroll direction

### LINKED ISSUE #2228: Shift+mouse wheel should scroll sideways
I think it's fairly standard for applications to scroll sideways with shift + scroll wheel. I would like to have this behavior for the ScrollView.

Related to #1280 I suppose.

### PR DESCRIPTION
**Pull request by ogoffart**
Closes #2228

### CODE REVIEW

**Comment by mgunyho** [comment]
Just a small comment: to me "reverse" sounds like (x, y) -> (-x, -y) instead of (y, x). But this is not in any public facing docs, so not a big deal.

Also, I haven't used mac-like touchpads that much, does shift+sideways scroll really slide up/down in other frameworks/apps?

**Review by tronical — APPROVED** [review]

**Comment by ogoffart** [comment]
> Also, I haven't used mac-like touchpads that much, does shift+sideways scroll really slide up/down in other frameworks/apps?

I actually don't know.  @tronical do you know?

File: tests/cases/elements/flickable.slint [hunk 1]
```diff
@@ -151,5 +151,21 @@ assert!((instance.get_offset_y() - 55.).abs() < 5.);
...
 
+```rust
+// Test wheel events
+use slint::{LogicalPosition, platform::{WindowEvent, Key} };
+let instance = TestCase::new().unwrap();
+instance.window().dispatch_event(WindowEvent::PointerScrolled { position: LogicalPosition::new(175.0, 175.0), delta_x: -30.0, delta_y: -50.0 });
+assert_eq!(instance.get_offset_x(), 30.);
+assert_eq!(instance.get_offset_y(), 50.);
+
+// When shift is pressed, it reverse the direction
```
Comment (ogoffart) [inline tests/cases/elements/flickable.slint:162]:
```suggestion
// When shift is pressed, it invert the direction
```

**Comment by tronical** [comment]
> > Also, I haven't used mac-like touchpads that much, does shift+sideways scroll really slide up/down in other frameworks/apps?
> 
> I actually don't know. @tronical do you know?

For me pressing shift while using the touch pad on macOS does not change the scrolling direction.

This patch changes the behavior when using a mouse wheel. For me, on macOS, the scroll direction changes when using the mouse wheel (external mouse) and pressing shift.

**Comment by mgunyho** [comment]
Okay yeah, I would expect shift to affect the wheel only. I was just a bit confused that x is also mapped to y. I guess normally the `WindowEvent::PointerScrolled` will only have an x or y component, not both.

**Comment by ogoffart** [comment]
This changes indeed invert the direction of the scroll event, which comes for, i think, both the trackpad and the wheel. I don't think we can differentiate at this point.
Do mac os already invert the direction of the wheel when pressing shift? Then this patch would invert it twice and that'd be wrong. And we'd need to ignore shift on mac.

Otherwise I think we can merge that even if it also invert the track pad. Maybe later we can attempt to differentiate.

**Comment by tronical** [comment]
> This changes indeed invert the direction of the scroll event, which comes for, i think, both the trackpad and the wheel. I don't think we can differentiate at this point. 

One difference between a wheel event on a touch pad vs. one from a real mouse is what (at least on my machine) the sequences differ:

1. The mouse wheel generates `WindowEvent::MouseWheel { phase: window::event::TouchPhaseMoved }`
2. The touch pad first generates `TouchPhase::Started`, then `::Moved`, and finally `Ended`.

So if we receive a wheel with `Started` then we know that we shouldn't invert, until we receive `Ended`?

> Do mac os already invert the direction of the wheel when pressing shift? Then this patch would invert it twice and that'd be wrong. And we'd need to ignore shift on Mac.

Yes, that is what happens. Without this patch, the shift key inverts the direction as expectedly, as it's done by macOS. With this patch it's inverted twice.

**Comment by ogoffart** [comment]
> So if we receive a wheel with Started then we know that we shouldn't invert, until we receive Ended?

But you're talking winit event,  we don't propagate these yet to the slint core.

Anyway we don't need to invert anything on mac apparently, since the system already does it.

> Yes, that is what happens. Without this patch, the shift key inverts the direction as expectedly, as it's done by macOS. With this patch it's inverted twice.

Then I see two possibilities:

1. keep this patch, but add a `cfg!(target_os = "macos")` to the condition
2. scrape this patch and instead handle that in the winit backend or other backend. So it would be the backend responsability to send these inverted event to Slint core.

@tronical: What do you think?

1. is easier and kind of fix the problem for @mgunyho use case, i believe
2. means that when we forward the scroll event to slint (#2228) the wheel event will be inverted with Slint if shift is pressed, no matter which widget or if shift should have another meaning there. But it will be consistent on all OS although this is not consistent with other toolkits on the other OS, i believe.

**Comment by tronical** [comment]
> > So if we receive a wheel with Started then we know that we shouldn't invert, until we receive Ended?
> 
> But you're talking winit event, we don't propagate these yet to the slint core.

Good point...

> Anyway we don't need to invert anything on mac apparently, since the system already does it.
> 
> > Yes, that is what happens. Without this patch, the shift key inverts the direction as expectedly, as it's done by macOS. With this patch it's inverted twice.
> 
> Then I see two possibilities:
> 
> 1. keep this patch, but add a `cfg!(target_os = "macos")` to the condition
> 2. scrape this patch and instead handle that in the winit backend or other backend. So it would be the backend responsability to send these inverted event to Slint core.
> 
> @tronical: What do you think?
>
> 1. is easier and kind of fix the problem for @mgunyho use case, i believe
> 2. means that when we forward the scroll event to slint ([Shift+mouse wheel should scroll sideways #2228](https://github.com/slint-ui/slint/issues/2228)) the wheel event will be inverted with Slint if shift is pressed, no matter which widget or if shift should have another meaning there. But it will be consistent on all OS although this is not consistent with other toolkits on the other OS, i believe.

I think 1. is fine for now. Perhaps we could implement 2. later by means of providing an actual scroll event as event type, besides "Wheel" (or pan gesture, more specifically).

**Comment by ogoffart** [comment]
Ok, i implemented 1.

>  later by means of providing an actual scroll event as event type, besides "Wheel" (or pan gesture, more specifically).

We only have "scroll" events (Wheel is an internal name, the public name in the platform::WindowEvent  is PointerScrolled   )

**Comment by tronical** [comment]
> Ok, i implemented 1.
> 
> > later by means of providing an actual scroll event as event type, besides "Wheel" (or pan gesture, more specifically).
> 
> We only have "scroll" events (Wheel is an internal name, the public name in the platform::WindowEvent is PointerScrolled )

Right, then "pan gesture" might make more sense (along with magnify, etc.)

**Review by tronical — APPROVED** [review]

**Pull request merged**